docs/api_overview.md

```diff
@@ -0,0 +1,472 @@
+  [WasmAccessLog](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/access_loggers/wasm/v3/wasm.proto#extensions-access-loggers-wasm-v3-wasmaccesslog)
+  extension).
+* `onDelete`: called after the plugin has completed all processing related to
+  the stream, as indication to release resources.
+
+`Context` instances also receive callbacks corresponding to stream events. For
+HTTP or gRPC streams, these are:
+
+* `onRequestHeaders`: called when HTTP or gRPC request headers are received.
+* `onRequestBody`: called when HTTP or gRPC request body data is received.
+* `onRequestTrailers`: called when HTTP or gRPC request trailers are received.
+* `onResponseHeaders`: called when HTTP or gRPC response headers are received.
+* `onResponseBody`: called when HTTP or gRPC response body data is received.
+* `onResponseTrailers`: called when HTTP or gRPC response trailers are received.
+
+For TCP streams, `Context` instances may receive the following callbacks:
+
+* `onNewConnection`: called when a new connection is established.
+* `onDownstreamData`: called when a new chunk of data is received from
+  downstream over a connection.
+* `onUpstreamData`: called when a new chunk of data is received from upstream
+  over a connection.
+
+Callback methods return status enums that indicate whether and how the host
+should continue to process the stream. Status enum meanings are specified in the
+doc comments in [proxy_wasm_enums.h]. Callbacks can also generate an immediate
+local response to an HTTP or gRPC request using the `sendLocalResponse`
+hostcall.
+
+For API details, see doc comments for the `Context` class in [proxy_wasm_api.h].
+
+## Buffers
+
+Many plugin operations require passing data buffers between the host and the
+plugin. Because this incurs some expense in copying data into or out of the Wasm
+VM's linear address space, data is only fetched or passed when explicitly
+requested via one of the following hostcalls:
+
+* `getBufferBytes`: retrieves bytes from a given buffer.
+* `getBufferStatus`: returns whether a given buffer is available, and if
+  so, what its size is
+* `setBuffer`: writes bytes into a given buffer.
+
+With each of these calls, buffer data is held by an instance of the `WasmData`
+class, which represents a span of bytes. The buffer to use as a source or
+destination of data is indicated by the `WasmBufferType` enum. For example, the
+`WasmBufferType::HttpResponseBody` enum value indicates the buffer used for
+passing or receiving HTTP response body data.
+
+Certain buffer types are only accessible during certain callbacks. For example,
+the `WasmBufferType::HttpResponseBody` buffer is only accessible during
+`Context::onResponseBody` and `Context::onLog` callbacks. See doc comments for
+individual callbacks in [proxy_wasm_api.h] for which buffers are accessible in
+each callback. Documentation for the `WasmData` class is also in
+[proxy_wasm_api.h].
+
+## Handling HTTP and gRPC requests
```

Comment:
This doesn't cover BODY chunks, so maybe call this "Handling HTTP headers and trailers" ?

Reply:
Done